model-optimizer/extensions/middle/InsertSelect.py

```diff
@@ -36,38 +48,63 @@ def run_before(self):
         return [ReplaceSpliceNodePattern]
 
     @staticmethod
-    def pattern():
-        return dict(
-            nodes=[('op', dict(op='Assign'))],
-            edges=[])
+    def calculate_frame_time(graph: Graph):
+        # there are either one or two inputs in Kaldi. Only main input can change delay in network.
+        # Usually ivector input has name 'ivector'.
+        inputs = graph.get_op_nodes(op='Parameter')
+        if len(inputs) == 1:
+            inp_name = inputs[0].name
+        elif len(inputs) == 2:
+            if inputs[0].name == 'ivector':
+                inp_name = inputs[1].name
+            elif inputs[1].name == 'ivector':
+                inp_name = inputs[0].name
+            else:
+                raise Error("There are 2 inputs for Kaldi model but we can't find out which one is ivector. " +
+                            "Use name \'ivector\' for according input")
+        else:
+            raise Error("There are {} inputs for Kaldi model but we expect only 1 or 2".format(len(inputs)))
+
+        # sort nodes to calculate delays
+        nodes = list(bfs_search(graph, [inp_name]))
+        nx.set_node_attributes(G=graph, name='frame_time', values=-1)
+
+        for n in nodes:
+            node = Node(graph, n)
+
+            # just ignore data nodes
+            if node.kind != 'op':
+                continue
+
+            # calculate frame_time (delay) that was not calculated
+            if node.frame_time < 0:
+                # Splice increases frame delay
+                if node.kind == 'op' and node.op == "Splice":
```

Comment:
no need to check for node kind since we have already filtered them out above. The same comment for line 85